modules/compiler/vecz/source/vector_target_info.cpp

```
@@ -46,23 +46,56 @@ Value *applyEVLToMask(IRBuilder<> &B, Value *EVL, Value *Mask) {
   return Mask;
 }
 
-bool isLegalMaskedLoad(const TargetTransformInfo &TTI, Type *Ty,
-                       unsigned Alignment) {
+template <typename TargetTransformInfo>
+auto isLegalMaskedLoadImpl(const TargetTransformInfo &TTI, Type *Ty,
+                           unsigned Alignment, unsigned)
+    -> decltype(TTI.isLegalMaskedLoad(Ty, Align(Alignment))) {
   return TTI.isLegalMaskedLoad(Ty, Align(Alignment));
 }
 
-bool isLegalMaskedStore(const TargetTransformInfo &TTI, Type *Ty,
-                        unsigned Alignment) {
+template <typename TargetTransformInfo>
+auto isLegalMaskedStoreImpl(const TargetTransformInfo &TTI, Type *Ty,
+                            unsigned Alignment, unsigned)
+    -> decltype(TTI.isLegalMaskedStore(Ty, Align(Alignment))) {
   return TTI.isLegalMaskedStore(Ty, Align(Alignment));
 }
 
+#if LLVM_VERSION_GREATER_EQUAL(21, 0)
+// TODO: Make this depend only on LLVM version once we do not have to remain
+// compatible with slightly older LLVM 21 snapshots.
+
+template <typename TargetTransformInfo>
+auto isLegalMaskedLoadImpl(const TargetTransformInfo &TTI, Type *Ty,
+                           unsigned Alignment, unsigned AddrSpace)
+    -> decltype(TTI.isLegalMaskedLoad(Ty, Align(Alignment), AddrSpace)) {
+  return TTI.isLegalMaskedLoad(Ty, Align(Alignment), AddrSpace);
+}
+
+template <typename TargetTransformInfo>
+auto isLegalMaskedStoreImpl(const TargetTransformInfo &TTI, Type *Ty,
+                            unsigned Alignment, unsigned AddrSpace)
+    -> decltype(TTI.isLegalMaskedStore(Ty, Align(Alignment), AddrSpace)) {
+  return TTI.isLegalMaskedStore(Ty, Align(Alignment), AddrSpace);
+}
+#endif
+
+bool isLegalMaskedLoad(const TargetTransformInfo &TTI, Type *Ty,
+                       unsigned Alignment, unsigned AddrSpace) {
+  return isLegalMaskedLoadImpl(TTI, Ty, Alignment, AddrSpace);
```

How does this work for < 21 - is it not undefined?
Also should this kind of logic be in multi_llvm?

---

`isLegalMaskedLoadImpl` is overloaded; the version that works for LLVM 19 and 20, and older snapshots of LLVM 21, is always defined but silently ignored on newer LLVM 21.

This could be in `multi_llvm` but we've generally only done that when we want something re-used in multiple places, have we not?

---

ah fair enough I couldn't see in the change that it existed elsewhere.
multi_llvm was just a thought, but it's where we've done this new method of supporting functions in this way, so it might benefit from consistency in one place.
Up to you though. I approve regardless.

---

Alright, done.